### PR TITLE
[SPIKE] Atomic baked in styles

### PR DESCRIPTION
**Pull request by madou**
This is a spike which refactors the internals to start as atomic declarations. It's rough - and we'll probably throw it away and re-write it afterwards. For now we're just using it to gather data. #272 

It does work though!

Before (overrides cancelled):

image

After (overrides persisted):

image

## Observations

- Bundle size gets bigger
- Gzip keeps bundle size in a reasonable state
- Atomic declarations naturally solve the overriding problem

## Todo

- [x] Create a complex example
- [x] Get bundle size (mingzip) before
- [x] Get bundle size (mingzip) after
- [x] Investigate alternate hashing solutions
- [x] Investigate ensuring atomic styles within a single module
- [ ] Ordering of the runtime stylesheet to ensure pseduo selector order

### CODE REVIEW

**Comment by pgmanutd** [comment]
So what I was saying, we can get numbers around how many unique atomic classes we have and how much bundle size they are increasing.

**For eg.** 
Suppose we have 50 unique atomic classes and 50 components are using min. 5 (and max 10) of them at the same time. This might add additional bundle size of suppose `~2KB`. And in future we might have 200-300 unique classes and 1000 of components. So it might give us rough estimate of what we can expect in future right. I know this sounds weird, but hehe 🤷‍♂️ 😛

And what if we can combine similar classes and create one for them? I know this deviates us from being atomified, but we can get smart and reduce bundlesize if we see similar classes are being used many times (above a minimum threshold) by components. And we can also let other classes being atomified if they are not used many times.

**For eg.**
```typescript
expect(actual.sheets.join('\n')).toMatchInlineSnapshot(`
      ".cc-v5rl1c-h2mmj6{position:relative}
      .cc-cy1bws-1dk0nxm{text-transform:capitalize}
      .cc-1aqc269-16l8ovn:after{content:\\"›\\"}
      .cc-p8hyfu-stnw88:after{position:absolute}
      .cc-1982bvg-lgv52v:after{right:-2rem}"
    `);
```   

will become 

```typescript
expect(actual.sheets.join('\n')).toMatchInlineSnapshot(`
      ".cc-abcde-fghijk{position:relative;text-transform:capitalize}
      .cc-abcde-fghijk:after{content:\\"›\\";position:absolute;right:-2rem}
    `);
```

**Comment by madou** [comment]
Running with these settings `{ nonce: '"k0Mp1lEd"', minify: true }`:

```
npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
npx babel .tmp/compiled.js --presets minify --out-file .tmp/compiled.min.js
npx gzip-size .tmp/compiled.min.js
```

## Atomic baked

```
➜  compiled git:(master) ✗ npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
➜  compiled git:(master) ✗ npx babel .tmp/compiled.js --presets minify --out-file .tmp/compiled.min.js
➜  compiled git:(atomic-baked-in-styles) ✗ npx gzip-size .tmp/compiled.min.js
3.55 kB
```

## Previous baked

```
➜  compiled git:(master) ✗ npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
➜  compiled git:(master) ✗ npx babel .tmp/compiled.js --presets minify --out-file .tmp/compiled.min.js
➜  compiled git:(master) ✗ npx gzip-size .tmp/compiled.min.js
2.9 kB
```

## Difference

```
0.65kB
```

**Comment by madou** [comment]
Current extract spike:

```
➜  compiled git:(atomic-baked-in-styles) ✗ npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
➜  compiled git:(atomic-baked-in-styles) ✗ npx babel .tmp/compiled.js --presets minify --out-file .tmp/compiled.min.js
➜  compiled git:(atomic-baked-in-styles) ✗ npx gzip-size .tmp/compiled.min.js
3.86 kB
```

It's actually bigger. Let me keep testing.

**Comment by madou** [comment]
Update, after fixing some things and turning mangling on I get it a smaller bundle size (but still not as small as before):

```
➜  compiled git:(atomic-baked-in-styles) ✗ npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
➜  compiled git:(atomic-baked-in-styles) ✗ npx minify .tmp/compiled.js --out-file .tmp/compiled.min.js --mangle.topLevel
➜  compiled git:(atomic-baked-in-styles) ✗ npx gzip-size .tmp/compiled.min.js
3.37 kB
```

Difference: `0.47kB`

**Comment by madou** [comment]
Reducing the hash down we get this:

```
➜  compiled git:(atomic-baked-in-styles) ✗ npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
➜  compiled git:(atomic-baked-in-styles) ✗ npx minify .tmp/compiled.js --out-file .tmp/compiled.min.js --mangle.topLevel
➜  compiled git:(atomic-baked-in-styles) ✗ npx gzip-size .tmp/compiled.min.js
3.33 kB
```

Difference: `0.43kB`

**Comment by madou** [comment]
Reducing hashes even further we get:

```
➜  compiled git:(atomic-baked-in-styles) ✗ npx babel examples/complex-example.tsx --out-file .tmp/compiled.js
➜  compiled git:(atomic-baked-in-styles) ✗ npx minify .tmp/compiled.js --out-file .tmp/compiled.min.js --mangle.topLevel
➜  compiled git:(atomic-baked-in-styles) ✗ npx gzip-size .tmp/compiled.min.js
3.15 kB
```

Difference: `0.25kB`

I think this is in the realm of "okay". I'll push my changes up and tomorrow morning I'll write up a Confluence page with next steps.

**Comment by madou** [comment]
TLDR the generated file looks like:

```
const a = ".1xum1fqe{padding:8px 12px}";
const b = ".1ylxu67f{color:#fff}";
const c = ".1k611j28{background-color:transparent}";
const d = ".pruiglyw{border:none}";
const e = ".14w9dlk8{font-size:14px}";
const f = ".1y9u1l7b{border-radius:3px}";
const g = ".ysyxz9vh:hover{background:rgba(116,118,121,0.6)}";

const Button = ({
  children
}) => /*#__PURE__*/React.createElement(_core.CC, null, /*#__PURE__*/React.createElement(_core.CS, {
  nonce: "k0Mp1lEd"
}, [a, b, c, d, e, f, g]), /*#__PURE__*/React.createElement("button", {
  className: (0, _core.ax)("1xum1fqe 1ylxu67f 1k611j28 pruiglyw 14w9dlk8 1y9u1l7b ysyxz9vh 1k9tz9vh")
}, children));
```

Multiple declarations get reused in the same file.

File: packages/css/src/plugins/atomicify.tsx [hunk 1]
```diff
@@ -28,8 +28,8 @@ const normalizeSelector = (str?: string) => {
 
 const atomicifyDecl = (node: Declaration, opts: Opts = {}) => {
   const normalizedSelector = normalizeSelector(opts.parentSelector);
-  const group = hash(`${join(opts.parentAtRule, opts.parentSelector)}-${node.prop}`);
-  const className = `cc-${group}-${hash(node.value)}`;
+  const group = hash(`${join(opts.parentAtRule, opts.parentSelector)}${node.prop}`).slice(0, 4);
+  const className = `${group}${hash(node.value).slice(0, 4)}`;
```
Comment (CraigCav) [inline packages/css/src/plugins/atomicify.tsx:32]: It looks like this change would  break the ax runtime - which splits the class name on the `-` in order to identify the group (for dedupe/handling overridden styles).

File: packages/babel-plugin/src/css-prop/__tests__/object-literal.test.tsx [hunk 2]
```diff
@@ -32,17 +32,23 @@ describe('css prop object literal', () => {
     expect(actual).toInclude('style={{"--var-hash-test":cl}}');
   });
 
-  it('should inline constant variable', () => {
+  it.only('should inline constant variable', () => {
```
Comment (pgmanutd) [inline packages/babel-plugin/src/css-prop/__tests__/object-literal.test.tsx:35]: oops 😛

File: packages/babel-plugin/src/css-prop/__tests__/string-literal.test.tsx [hunk 3]
```diff
@@ -30,7 +30,7 @@ describe('css prop string literal', () => {
     expect(actual).toInclude('style={{"--var-hash-test":(fontSize||"")+"px"}}');
   });
 
-  it('should persist suffix of constant value', () => {
+  it.only('should persist suffix of constant value', () => {
```
Comment (pgmanutd) [inline packages/babel-plugin/src/css-prop/__tests__/string-literal.test.tsx:33]: hehe oops 😛

**Review by pgmanutd — COMMENTED** [review]
Will review in morning. Very tired 😞. Already looked once, will review in detail tmr morning.

**Comment by madou** [comment]
no rush @pgmanutd

File: packages/css/src/plugins/atomicify.tsx (hunk 1 above)
Comment (madou) [reply packages/css/src/plugins/atomicify.tsx:32]: good catch! yeah you're right - we'd be changing `ax` to get the first 4 characters instead

File: packages/babel-plugin/src/css-prop/__tests__/string-literal.test.tsx (hunk 3 above)
Comment (madou) [reply packages/babel-plugin/src/css-prop/__tests__/string-literal.test.tsx:33]: its a spike! DW 😂

**Comment by alexreardon** [comment]
I am not completely sold on this yet (i haven't thought it through)

That said - i think compiled gives us the flexibility to change our mind on this approach as we go

**Comment by madou** [comment]
let's catch up today and talk through it alex

**Comment by madou** [comment]
@alexreardon so I've been doing some testing this morning - gzip does quite well at reducing size when multiple duplicate atomic declarations are inside a single bundle.

**Original**

This is the current ADS example.

```bash
➜  compiled git:(atomic-baked-in-styles) npx gzip-size .tmp/compiled.min.js
3.16 kB
```

**Duplicated declarations**

Takes the ADS example and duplicates the declarations.

```bash
➜  compiled git:(atomic-baked-in-styles) npx gzip-size .tmp/compiled.min.js
3.13 kB
```

_It's smaller_. This makes sense because of how gzip can optimize duplicated lengths of strings in a single file. I have to wonder if we took this to an entire app what would this look like.

---

When we talk around the alternate strategy of compiling overrides, so for example:

```
const StyledDiv = styled.div`
  color: red;
  font-size: 12px;
`;

const WrappedStyled = styled(StyledDiv)`
  color: blue;
`;
```

Which would then produce these styles for wrapped style:

```
.hash { color: blue; font-size: 12px }
```

It does paint a good story for using the less boilerplate strings compared to the atomic declarations approach. But - where can this break expectations, or make things harder for us?

1. Overriding styles on wrapping components
1. Overriding already baked components 
1. Conditionally applying styles

**Overriding styles on wrapping components**

This is more insidious when we consider this would naturally happen when using CSS prop, but it can happen for all APIs. 

```jsx
const Component = ({ className }) => <div css={{ fontSize: 12, className }} />
export const WrappedComponent = (props) => <Component {...props} />;
```
```jsx
import WrappedComponent from './component';

styled(WrappedComponent)`
  font-size: 20px;
`;
```

We now need to keep traversing until we _might_ find a component. It's a lot more traversal.

**Overriding already baked components** 

When considering consuming packages from NPM - they're already baked. We will need to traverse already transpiled code and try to analyze the defined styles. This will be more code path to handle these scenarios.

**Conditionally applying styles**

Take this code.

```jsx
const StyledDiv = styled.div`
  color: red;
  font-size: 12px;
`;

const WrappedStyled = () => {
  const [isBig] = useState(false);
  return <StyledDiv css={[isBig && { fontSize: 20 }]} />
};
```

Fundamentally this code can't work. We wouldn't be able to support this API.

---

Atomic declarations solve this problem quite naturally - with the biggest bonus being all the above cases now work as expected. But - as we've seen - at the cost of bundle size increasing. I think we need more data to understand the true ramifications of this - however:

1. gzipping the example with duplicated declarations has shown to be quite effective
1. the bundle size increase will be made irrelevant when we extract to CSS

Finally I don't think we should underestimate the benefits of _behaviour_ being the same when baked and extracted. We will inevitably find bugs or fundamental blockers in our implementation if they diverge (like this spike has been aiming to solve 😄).

We have flexibility and potential to optimize the further we go on - if we go down this path we can continually evaluate and adjust as needed.

File: packages/css/src/plugins/atomicify.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import { hash } from '@compiled/utils';
+
+interface Opts {
+  parentSelector?: string;
+  parentAtRule?: string;
+  callback?: (className: string) => void;
+}
+
+const join = (...str: (string | undefined)[]) => {
+  if (!str) {
```
Comment (pgmanutd) [inline packages/css/src/plugins/atomicify.tsx:11]: nit: Will this be undefined at any point of time since its a rest parameter.

File: packages/runtime/src/__tests__/ax.test.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import ax from '../ax';
+
+describe('ax', () => {
+  it('should remove duplicates', () => {
```
Comment (pgmanutd) [inline packages/runtime/src/__tests__/ax.test.tsx:4]: ahh! I was looking for this duplicate one. Nice ❤️

File: packages/runtime/src/ax.tsx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export default function ax(className: string) {
+  const found: any = {};
+
+  className.split(' ').forEach((cls) => {
+    const index = cls.charAt(0) === '_' ? 5 : 4;
```
Comment (pgmanutd) [inline packages/runtime/src/ax.tsx:5]: nit: 😛 5 and 4 looks like magic numbers.

**Review by pgmanutd — COMMENTED** [review]
@madou: you can ignore nits if we are not merging this into master. Anyways this is just a spike right. nice work ❤️.

**Comment by madou** [comment]
yeah exactly we will close this and break it down into smaller discreet chunks

- PostCSS atomic plugin
- atomic ax runtime + update ast builder to use ax runtime
- update style component to use not add based on stylesheets instead of hash
- update ast builder to use ax runtime
- hoist style declarations to the top of the module

**Comment by madou** [comment]
Spike completed. Follow up tasks to follow.
